services/dbaas/dbaas_client.go

```diff
@@ -19,102 +19,245 @@ package dbaas
-		logsClient:          controllerv1beta1.NewLogsAPIClient(con),
+func NewClient(dbaasControllerAPIAddress string) *Client {
+	c := &Client{
+		l:                         logrus.WithField("component", "dbaas.Client"),
+		connM:                     new(sync.RWMutex),
+		dbaasControllerAPIAddress: dbaasControllerAPIAddress,
+		wg:                        new(sync.WaitGroup),
+	}
+	return c
+}
+
+// Connect connects the client to dbaas-controller API.
+func (c *Client) Connect(ctx context.Context) error {
+	c.connM.Lock()
+	defer c.connM.Unlock()
+	c.l.Infof("Connecting to dbaas-controller API on %s.", c.dbaasControllerAPIAddress)
+	if c.conn != nil {
+		c.l.Warnf("Trying to connect to dbaas-controller API but connection is already up.")
+		return nil
+	}
+	backoffConfig := backoff.DefaultConfig
+	backoffConfig.MaxDelay = 10 * time.Second
+	opts := []grpc.DialOption{
+		grpc.WithBlock(), // Dial blocks, we do not connect in background.
+		grpc.WithInsecure(),
+		grpc.WithConnectParams(grpc.ConnectParams{Backoff: backoffConfig, MinConnectTimeout: 10 * time.Second}),
+		grpc.WithUserAgent("pmm-managed/" + version.Version),
+	}
+
+	conn, err := grpc.DialContext(ctx, c.dbaasControllerAPIAddress, opts...)
+	if err != nil {
+		return errors.Errorf("failed to connect to dbaas-controller API: %v", err)
 	}
+	c.conn = conn
+
+	c.kubernetesClient = controllerv1beta1.NewKubernetesClusterAPIClient(conn)
+	c.xtradbClusterClient = controllerv1beta1.NewXtraDBClusterAPIClient(conn)
+	c.psmdbClusterClient = controllerv1beta1.NewPSMDBClusterAPIClient(conn)
+	c.logsClient = controllerv1beta1.NewLogsAPIClient(conn)
+
+	c.l.Info("Connected to dbaas-controller API.")
+	return nil
+}
+
+// Disconnect disconnects the client from dbaas-controller API.
+func (c *Client) Disconnect() error {
+	c.connM.Lock()
+	defer c.connM.Unlock()
+	c.l.Info("Disconnecting from dbaas-controller API.")
+	c.wg.Wait()
+	if c.conn == nil {
+		c.l.Warnf("Trying to disconnect from dbaas-controller API but the connection is not up.")
+		return nil
+	}
+
+	if err := c.conn.Close(); err != nil {
+		return errors.Errorf("failed to close conn to dbaas-controller API: %v", err)
+	}
+	c.conn = nil
+	c.l.Info("Disconected from dbaas-controller API.")
+	return nil
 }
 
 // CheckKubernetesClusterConnection checks connection with kubernetes cluster.
 func (c *Client) CheckKubernetesClusterConnection(ctx context.Context, kubeConfig string) (*controllerv1beta1.CheckKubernetesClusterConnectionResponse, error) {
-	return c.kubernetesClient.CheckKubernetesClusterConnection(ctx, &controllerv1beta1.CheckKubernetesClusterConnectionRequest{
-		KubeAuth: &controllerv1beta1.KubeAuth{Kubeconfig: kubeConfig},
-	})
+	c.connM.RLock()
+	defer c.connM.RUnlock()
+	c.wg.Add(1)
+
+	in := &controllerv1beta1.CheckKubernetesClusterConnectionRequest{
+		KubeAuth: &controllerv1beta1.KubeAuth{
+			Kubeconfig: kubeConfig,
+		},
+	}
+	out, err := c.kubernetesClient.CheckKubernetesClusterConnection(ctx, in)
```

Comment:
let's check if connection is exists.

Reply:
Do you mean `c.conn`? If it does not exist, it means dbaas is disabled and `Connect` was not called. So I think there is no need to check it.

Reply:
We did not do any checking of the result of the rpc call previously, not sure how it would help us now, if that is what you are requesting.

Reply:
ok